### PR TITLE
add multi-arch

### LINKED ISSUE #451: Official arm64 build for the controller
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

As specified here, images on bintray should follow have an architecture appended to them:
https://github.com/Kong/kong/blob/master/CHANGELOG.md

Currently only amd64 is available.

Kong proxy does but the ingress controller does not. Currently, without rebuilding and uploading to another repository, this is stopping users deploying kong to Arm64.

### PR DESCRIPTION
**Pull request by gpabdo**
** Add arm64 build to official repo
** fixes #451

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,6 +47,16 @@ container:
     --build-arg REPO_INFO=${REPO_INFO} \
     -t ${IMAGE}:${TAG} .
 
+.PHONY: multi-arch
+multi-arch:
+	docker buildx build \
```
Comment (hbagdi) [inline Makefile:52]: I'm not familiar with docker. I've `19.03.12` installed and `docker buildx` is not a sub-command in my installation.

File: Makefile [hunk 2]
```diff
@@ -47,6 +47,16 @@ container:
...
+    --platform linux/amd64,linux/arm64 \
```
Comment (hbagdi) [inline Makefile:53]: Does this result in two different images with the same tag? If so, then how are they supposed to be hosted and pulled?

**Comment by hbagdi** [comment]
One blocker for this would be to figure out if Bintray supports multi-arch images or not.

File: Makefile (hunk 1 above)
Comment (gpabdo) [reply Makefile:52]: Good call, looks like this part is pretty easy as of Docker 19.03. https://github.com/Kong/kubernetes-ingress-controller/blob/8132400700df912278e2d9d2592f3c1a54648de4/.ci/build-main.sh#L6 

https://docs.docker.com/buildx/working-with-buildx/

File: Makefile (hunk 2 above)
Comment (gpabdo) [reply Makefile:53]: I am not 100% sure how it knows what architecture a container was built for... Here when I pull `ubuntu:latest` it just knows... https://hub.docker.com/_/ubuntu?tab=tags I hope maybe @lyndon160 can provide more insight here?

Comment (lyndon160) [reply Makefile:53]: This results in two images and one manifest. The manifest then points to each image. When a docker pull is made, it reads the manifest to pull the image that matches your architecture.

If multi-arch images are not supported then simply do the `buildx` command for arm64 and tag it something like kong:1.5-arm64

Unsure about Bintray, they don't have any documentation on this. Most of the popular registries offer multi-arch -- maybe test it locally first with an helloworld image? Again if this doesn't work, then tag images for each arch.

**Comment by hbagdi** [comment]
Noted. This likely requires some investigation from our end. We will get to it soon.
cc @mflendrich @rainest

**Comment by shaneutt** [comment]
The maintainers have discussed this recently, and there's some extra context in https://github.com/Kong/kubernetes-ingress-controller/issues/1019 which provides some per-requisites we need to sort out before we can commit to maintaining ARM builds. Further prioritization is needed and we'll keep this and the two related issues updated when we have new plans/progress.

**Comment by shaneutt** [comment]
Just wanted to bump this and note: we're still making progress towards multi-arch support, but it's slow going internally due to other higher priorities. We'll continue to keep track of this PR and update it as things progress.

**Comment by shaneutt** [comment]
Please see the following update regarding this PR and let us know (in the comments of the issue) whether you have any questions or concerns: https://github.com/Kong/kubernetes-ingress-controller/issues/451#issuecomment-856766224

**Pull request closed without merging**